File: app/src/main/java/pl/pola_app/helpers/SessionId.java
package pl.pola_app.helpers;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import java.util.UUID;

/**
 * Creates GUID.
 * It's stored in app's shared preferences so it's recreated when data are cleared.
 */
public class SessionId {
    private static final String PREF_SESSION_GUID = "session_guid";
    private static final Object lock = new Object();
    private final String sessionId;

    public static SessionId create(Context context) {
        return new SessionId(context);
    }

    private SessionId(Context context) {
        sessionId = init(context);
    }

    public String get() {
        return sessionId;
    }

    private static SharedPreferences getDefaultSharedPreferences(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }

    private static String init(Context context) {
        synchronized (lock) {
            SharedPreferences pref = getDefaultSharedPreferences(context);
            String sessionGuid = pref.getString(PREF_SESSION_GUID, null);

            if (sessionGuid == null) {
                sessionGuid = UUID.randomUUID().toString();
                final SharedPreferences.Editor editor = pref.edit();
                editor.putString(PREF_SESSION_GUID, sessionGuid);
                editor.apply();
            }
            return sessionGuid;
        }
    }
}

File: app/src/main/java/pl/pola_app/helpers/SessionId.kt
package pl.pola_app.helpers

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import java.util.*

/**
 * Creates GUID.
 * It's stored in app's shared preferences so it's recreated when data are cleared.
 */
open class SessionId(context: Context) {
    private val sessionId: String = init(context)
    fun get(): String {
        return sessionId
    }

    companion object {
        private const val PREF_SESSION_GUID = "session_guid"
        private val lock = Any()
        fun create(context: Context): SessionId {
            return SessionId(context)
        }

        private fun getDefaultSharedPreferences(context: Context): SharedPreferences {
            return PreferenceManager.getDefaultSharedPreferences(context)
        }

        private fun init(context: Context): String {
            synchronized(lock) {
                val pref =
                    getDefaultSharedPreferences(context)
                var sessionGuid =
                    pref.getString(PREF_SESSION_GUID, null)
                if (sessionGuid == null) {
                    sessionGuid = UUID.randomUUID().toString()
                    val editor = pref.edit()
                    editor.putString(PREF_SESSION_GUID, sessionGuid)
                    editor.apply()
                }
                return sessionGuid
            }
        }
    }

}